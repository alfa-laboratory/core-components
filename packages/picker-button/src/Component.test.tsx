import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';

import { PickerButton } from './index';

const testId = 'picker-button-test-id';
const className = 'pickerButtonClassName';
const label = 'some label';
const options = [
    { key: '1', content: 'Neptunium' },
    { key: '2', content: 'Plutonium' },
    { key: '3', content: 'Americium' },
    { key: '4', content: 'Curium' },
    { key: '5', content: 'Berkelium' },
    { key: '6', content: 'Californium' },
    { key: '7', content: 'Einsteinium' },
    { key: '8', content: 'Fermium' },
    { key: '9', content: 'Mendelevium' },
    { key: '10', content: 'Nobelium' },
    { key: '11', content: 'Lawrencium' },
    { key: '12', content: 'Rutherfordium' },
    { key: '13', content: 'Dubnium' },
    { key: '14', content: 'Seaborgium' },
    { key: '15', content: 'Bohrium' },
];

const clickPickerButton = () => {
    const button = document.querySelector('button') as HTMLButtonElement;
    fireEvent.click(button);
};

describe('Snapshots tests', () => {
    it('should display correctly', () => {
        const { container } = render(<PickerButton label={label} options={options} />);

        expect(container).toMatchSnapshot();
    });

    it('should display opened correctly', async () => {
        const { container } = render(<PickerButton options={options} />);

        clickPickerButton();

        await waitFor(() => {
            expect(container).toMatchSnapshot();
        });
    });
});

describe('Attributes tests', () => {
    it('should set data-test-id attribute', async () => {
        render(<PickerButton className={className} dataTestId={testId} options={options} />);

        const container = document.querySelector(`.${className}`);

        const testIdAttr = container?.getAttribute('data-test-id');

        expect(testIdAttr).toBe(testId);
    });

    it('should have disabled attribute', async () => {
        render(<PickerButton disabled={true} options={options} />);

        const button = document.querySelector('button');

        expect(button).toHaveAttribute('disabled');
    });
});

describe('Render tests', () => {
    it('should unmount without errors', async () => {
        const { unmount } = render(<PickerButton options={options} />);

        expect(unmount).not.toThrowError();
    });

    it('should have loading class', async () => {
        render(<PickerButton options={options} loading={true} />);
        const button = document.querySelector('button');

        expect(button).toHaveClass('loading');
    });

    it('should have primary class', async () => {
        const view = 'primary';
        render(<PickerButton options={options} view={view} />);
        const button = document.querySelector('button');

        expect(button).toHaveClass(view);
    });

    it('should have m class, icon size = 24, secondary class by default', async () => {
        const iconSize = '24';
        const { getByTestId } = render(<PickerButton options={options} />);
        const button = document.querySelector('button');
        const icon = getByTestId('picker-button-icon');

        expect(button).toHaveClass('m');
        expect(button).toHaveClass('secondary');
        expect(icon.getAttribute('width')).toBe(iconSize);
        expect(icon.getAttribute('height')).toBe(iconSize);
    });

    it('should have xs class and small icon', async () => {
        const size = 'xs';
        const iconSize = '18';

        const { getByTestId } = render(<PickerButton options={options} size={size} />);
        const button = document.querySelector('button');
        const icon = getByTestId('picker-button-icon');

        expect(button).toHaveClass(size);
        expect(icon.getAttribute('width')).toBe(iconSize);
        expect(icon.getAttribute('height')).toBe(iconSize);
    });

    it('should have open class if opened', async () => {
        render(<PickerButton label={label} options={options} />);

        clickPickerButton();

        await waitFor(() => {
            const iconContainer = document.querySelector('.iconContainer');
            expect(iconContainer).toHaveClass('open');
        });
    });

    it('should have options if opened', async () => {
        render(<PickerButton label={label} options={options} />);

        clickPickerButton();

        await waitFor(() => {
            const renderedOptions = document.querySelectorAll('.option');
            expect(renderedOptions).toHaveLength(options.length);
        });
    });
});
