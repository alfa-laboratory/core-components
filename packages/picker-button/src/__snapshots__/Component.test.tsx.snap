// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshots tests should display correctly 1`] = `
<div>
  <div
    aria-expanded="false"
    aria-haspopup="listbox"
    aria-owns="downshift-1-menu"
    class="component container"
    role="combobox"
    tabindex="-1"
  >
    <button
      aria-controls="downshift-1-menu"
      aria-labelledby="downshift-1-label"
      class="component secondary m block component"
      id="downshift-0-input"
      tabindex="0"
      type="button"
    >
      <span
        class="text"
      >
        <span
          for="downshift-0-input"
          id="downshift-1-label"
        >
          some label
        </span>
      </span>
      <span
        class="addons"
      >
        <span
          class="iconContainer"
        >
          <svg
            data-test-id="picker-button-icon"
            fill="currentColor"
            focusable="false"
            height="24"
            role="img"
            viewBox="0 0 24 24"
            width="24"
          >
            <path
              d="M16.074 9.646l.708.708-5.068 5.068-5.068-5.068.708-.708 4.36 4.362z"
            />
          </svg>
        </span>
      </span>
    </button>
  </div>
</div>
`;

exports[`Snapshots tests should display opened correctly 1`] = `
<div>
  <div
    aria-expanded="true"
    aria-haspopup="listbox"
    aria-owns="downshift-3-menu"
    class="component container"
    role="combobox"
    tabindex="-1"
  >
    <button
      aria-controls="downshift-3-menu"
      aria-labelledby="downshift-3-label"
      class="component secondary m block iconOnly component"
      id="downshift-2-input"
      tabindex="0"
      type="button"
    >
      <span
        class="addons"
      >
        <span
          class="iconContainer open"
        >
          <svg
            data-test-id="picker-button-icon"
            fill="currentColor"
            focusable="false"
            height="24"
            role="img"
            viewBox="0 0 24 24"
            width="24"
          >
            <path
              d="M16.074 9.646l.708.708-5.068 5.068-5.068-5.068.708-.708 4.36 4.362z"
            />
          </svg>
        </span>
      </span>
    </button>
  </div>
</div>
`;

exports[`Snapshots tests should display opened correctly 2`] = `
<div
  alfa-portal-container=""
>
  <div
    class="component"
    data-popper-escaped="true"
    data-popper-placement="bottom-start"
    data-popper-reference-hidden="true"
    style="z-index: 50; position: absolute; left: 0px; top: 0px; transform: translate(0px, 0px);"
  >
    <div
      class="inner popoverInner"
    >
      <div
        aria-labelledby="downshift-3-label"
        class="optionsList"
        id="downshift-3-menu"
        role="listbox"
        style="min-width: 0px;"
      >
        <div
          class="optionsList m"
          style="height: 0px;"
        >
          <div
            aria-selected="true"
            class="option m highlighted"
            id="downshift-3-item-0"
            role="option"
          >
            <span
              class="checkmark"
            />
            <div
              class="content"
            >
              Neptunium
            </div>
          </div>
          <div
            aria-selected="false"
            class="option m"
            id="downshift-3-item-1"
            role="option"
          >
            <span
              class="checkmark"
            />
            <div
              class="content"
            >
              Plutonium
            </div>
          </div>
          <div
            aria-selected="false"
            class="option m"
            id="downshift-3-item-2"
            role="option"
          >
            <span
              class="checkmark"
            />
            <div
              class="content"
            >
              Americium
            </div>
          </div>
          <div
            aria-selected="false"
            class="option m"
            id="downshift-3-item-3"
            role="option"
          >
            <span
              class="checkmark"
            />
            <div
              class="content"
            >
              Curium
            </div>
          </div>
          <div
            aria-selected="false"
            class="option m"
            id="downshift-3-item-4"
            role="option"
          >
            <span
              class="checkmark"
            />
            <div
              class="content"
            >
              Berkelium
            </div>
          </div>
          <div
            aria-selected="false"
            class="option m"
            id="downshift-3-item-5"
            role="option"
          >
            <span
              class="checkmark"
            />
            <div
              class="content"
            >
              Californium
            </div>
          </div>
          <div
            aria-selected="false"
            class="option m"
            id="downshift-3-item-6"
            role="option"
          >
            <span
              class="checkmark"
            />
            <div
              class="content"
            >
              Einsteinium
            </div>
          </div>
          <div
            aria-selected="false"
            class="option m"
            id="downshift-3-item-7"
            role="option"
          >
            <span
              class="checkmark"
            />
            <div
              class="content"
            >
              Fermium
            </div>
          </div>
          <div
            aria-selected="false"
            class="option m"
            id="downshift-3-item-8"
            role="option"
          >
            <span
              class="checkmark"
            />
            <div
              class="content"
            >
              Mendelevium
            </div>
          </div>
          <div
            aria-selected="false"
            class="option m"
            id="downshift-3-item-9"
            role="option"
          >
            <span
              class="checkmark"
            />
            <div
              class="content"
            >
              Nobelium
            </div>
          </div>
          <div
            aria-selected="false"
            class="option m"
            id="downshift-3-item-10"
            role="option"
          >
            <span
              class="checkmark"
            />
            <div
              class="content"
            >
              Lawrencium
            </div>
          </div>
          <div
            aria-selected="false"
            class="option m"
            id="downshift-3-item-11"
            role="option"
          >
            <span
              class="checkmark"
            />
            <div
              class="content"
            >
              Rutherfordium
            </div>
          </div>
          <div
            aria-selected="false"
            class="option m"
            id="downshift-3-item-12"
            role="option"
          >
            <span
              class="checkmark"
            />
            <div
              class="content"
            >
              Dubnium
            </div>
          </div>
          <div
            aria-selected="false"
            class="option m"
            id="downshift-3-item-13"
            role="option"
          >
            <span
              class="checkmark"
            />
            <div
              class="content"
            >
              Seaborgium
            </div>
          </div>
          <div
            aria-selected="false"
            class="option m"
            id="downshift-3-item-14"
            role="option"
          >
            <span
              class="checkmark"
            />
            <div
              class="content"
            >
              Bohrium
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
